include/deal.II/numerics/nonlinear.h

```diff
@@ -0,0 +1,582 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/exceptions.h>
+
+#include <deal.II/sundials/kinsol.h>
+
+#include <deal.II/trilinos/nox.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Selects a nonlinear solver by choosing between KINSOL or NOX.
+ * KINSOL and NOX are nonlinear solvers included in the SUNDIALS
+ * package and the TRILINOS package resp. IF no solver type is
+ * specified it will automaticlaly choose a solver based on what
+ * the use has installed, KINSOL having priority.
+ *
+ * By calling the @p solve function of this @p NonlinearSolverSelector, it selects the
+ * @p solve function of that @p Solver that was specified in the constructor
+ * of this class, similar to the @p SolverSelector class.
+ *
+ * <h3>Usage</h3>
+ * An example of code one would run with this class is the
+ * following:
+ * @code
+ * // Generate a @p NonlinearSolverSelector that uses @p KINSOL
+ * std::string solver = "kinsol";
+ *
+ * NonlinearSolverSelector<Vector<double>>::AdditionalData additional_data;
+ * additional_data.solver_type = solver;
+ *
+ * NonlinearSolverSelector<Vector<double>> nonlinear_solver(additional_data);
+ *
+ * // Functions that are required for solving a nonlinear problem,
+ * // which are utilized in both @p KINSOL and @p NOX.
+ * nonlinear_solver.reinit_vector = [&](Vector<double> &x) {...}
+ *
+ * nonlinear_solver.residual =
+ *           [&](const Vector<double> &evaluation_point,
+ *               Vector<double> &      residual) {...}
+ *
+ * nonlinear_solver.setup_jacobian =
+ *           [&](const Vector<double> &current_u,
+ *               const Vector<double> &current_f) {...}
+ *
+ * nonlinear_solver.solve_with_jacobian =
+ *           [&](const Vector<double> &rhs,
+ *               Vector<double> &      dst,
+ *               const double tolerance) {...}
+ *
+ * // Calling the @p solve function with an initial guess.
+ * nonlinear_solver.solve(current_solution);
+ * @endcode
+ */
+template <typename VectorType = Vector<double>>
+class NonlinearSolverSelector
```

Comment:
How generic is the vector type? Can you use a Trilinos vector with KINSOL, and a deal.II vector with NOX?

Reply:
We should add tests for different vector types (and maybe parallel), but maybe this can be a separate PR. Both support deal.II vectors (see the test in this PR), but I am not sure about parallel vectors.

Reply:
My last comment is incorrect, both NOX and KINSOL have MPI support, working on getting this class working with MPI